File: tests/test_scheduler.py
import asyncio
import pytest
from unittest.mock import AsyncMock
from scavenger.daemon.scheduler import PollScheduler
from scavenger.models import Profile


def make_profile(**overrides) -> Profile:
    return Profile(**{
        "id": "test", "name": "Test", "keywords": ["test"],
        "negative_keywords": [], "sources": ["ebay"],
        "poll_interval_sec": 60, "enabled": True, **overrides
    })


async def test_scheduler_starts_and_stops():
    s = PollScheduler()
    await s.start()
    assert s.running
    await s.stop()
    assert not s.running


async def test_add_profile_creates_job():
    s = PollScheduler()
    await s.start()
    s.add_profile(make_profile(), callback=AsyncMock())
    assert s.has_job("test")
    await s.stop()


async def test_disabled_profile_not_scheduled():
    s = PollScheduler()
    await s.start()
    s.add_profile(make_profile(enabled=False), callback=AsyncMock())
    assert not s.has_job("test")
    await s.stop()


async def test_remove_profile_removes_job():
    s = PollScheduler()
    await s.start()
    s.add_profile(make_profile(), callback=AsyncMock())
    s.remove_profile("test")
    assert not s.has_job("test")
    await s.stop()


async def test_trigger_now_calls_callback():
    s = PollScheduler()
    await s.start()
    mock_cb = AsyncMock(return_value=[])
    profile = make_profile()
    s.add_profile(profile, callback=mock_cb)
    await s.trigger_now("test")
    await asyncio.sleep(0.05)
    mock_cb.assert_called_once_with(profile)
    await s.stop()

File: rust/tests/test_scheduler.rs
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Arc;

use chrono::Utc;

use scavenger::daemon::scheduler::{PollCallback, PollScheduler};
use scavenger::models::Profile;

fn test_profile(id: &str, interval_sec: u64) -> Profile {
    Profile {
        id: id.to_string(),
        name: format!("Test {}", id),
        keywords: vec![],
        negative_keywords: vec![],
        sources: vec!["ebay".to_string()],
        poll_interval_sec: interval_sec,
        price_min: None,
        price_max: None,
        enabled: true,
    }
}

fn noop_callback() -> PollCallback {
    Arc::new(|_p| Box::pin(async { vec![] }))
}

fn counting_callback(counter: &Arc<AtomicU32>) -> PollCallback {
    let c = counter.clone();
    Arc::new(move |_p| {
        let c = c.clone();
        Box::pin(async move {
            c.fetch_add(1, Ordering::Relaxed);
            vec![]
        })
    })
}

#[tokio::test]
async fn add_profile_creates_job() {
    let scheduler = PollScheduler::new();
    scheduler.start().await;

    let profile = test_profile("test-1", 60);
    scheduler
        .add_profile(profile, noop_callback(), None)
        .await;

    assert!(scheduler.has_job("test-1").await);
    scheduler.stop().await;
}

#[tokio::test]
async fn remove_profile_cancels_job() {
    let scheduler = PollScheduler::new();
    scheduler.start().await;

    let profile = test_profile("test-2", 60);
    scheduler
        .add_profile(profile, noop_callback(), None)
        .await;
    assert!(scheduler.has_job("test-2").await);

    scheduler.remove_profile("test-2").await;
    assert!(!scheduler.has_job("test-2").await);
    scheduler.stop().await;
}

#[tokio::test]
async fn disabled_profile_not_added() {
    let scheduler = PollScheduler::new();
    scheduler.start().await;

    let mut profile = test_profile("disabled-1", 60);
    profile.enabled = false;
    scheduler
        .add_profile(profile, noop_callback(), None)
        .await;

    assert!(!scheduler.has_job("disabled-1").await);
    scheduler.stop().await;
}

#[tokio::test]
async fn trigger_now_fires_callback() {
    let counter = Arc::new(AtomicU32::new(0));
    let callback = counting_callback(&counter);

    let scheduler = PollScheduler::new();
    scheduler.start().await;

    // Use a long interval so the timer doesn't fire during the test
    let profile = test_profile("trigger-1", 3600);
    scheduler.add_profile(profile, callback, None).await;

    // Wait briefly for the initial tick to fire (interval fires immediately)
    tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
    let before_trigger = counter.load(Ordering::Relaxed);

    scheduler.trigger_now("trigger-1").await;
    tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;

    let after_trigger = counter.load(Ordering::Relaxed);
    assert!(
        after_trigger > before_trigger,
        "trigger_now should have fired the callback (before={}, after={})",
        before_trigger,
        after_trigger
    );
    scheduler.stop().await;
}

#[tokio::test]
async fn jitter_delays_when_recently_polled() {
    let counter = Arc::new(AtomicU32::new(0));
    let callback = counting_callback(&counter);

    let scheduler = PollScheduler::new();
    scheduler.start().await;

    let profile = test_profile("jitter-1", 100);
    // Last polled just now -- next run should be delayed by ~100s
    scheduler
        .add_profile(profile, callback, Some(Utc::now()))
        .await;

    // The callback should NOT fire in 300ms because the interval hasn't elapsed
    tokio::time::sleep(tokio::time::Duration::from_millis(300)).await;
    assert_eq!(
        counter.load(Ordering::Relaxed),
        0,
        "callback should not fire when last_polled is recent"
    );

    scheduler.stop().await;
}

#[tokio::test]
async fn stop_cancels_all_jobs() {
    let scheduler = PollScheduler::new();
    scheduler.start().await;

    for i in 0..3 {
        let profile = test_profile(&format!("stop-{}", i), 60);
        scheduler
            .add_profile(profile, noop_callback(), None)
            .await;
    }

    scheduler.stop().await;

    assert!(!scheduler.has_job("stop-0").await);
    assert!(!scheduler.has_job("stop-1").await);
    assert!(!scheduler.has_job("stop-2").await);
}
